pkg/driver/driver_azure.go

```diff
@@ -380,6 +379,33 @@ func (d *AzureDriver) getvms(machineID string, listOfVMs VMs) error {
 	return nil
 }
 
+// waitForDataDiskDetachment waits for data disks to be detached
+func (d *AzureDriver) waitForDataDiskDetachment(machineID string) error {
+	var cancel <-chan struct{}
+	d.setup()
+
+	vm, err := vmClient.Get(d.AzureMachineClass.Spec.ResourceGroup, machineID, compute.InstanceView)
+	if err != nil {
+		metrics.APIFailedRequestCount.With(prometheus.Labels{"provider": "azure", "service": "virtual_machine"}).Inc()
+		glog.Errorf("Failed to list VMs. Error Message - %s", err)
+		return err
+	}
+	metrics.APIRequestCount.With(prometheus.Labels{"provider": "azure", "service": "virtual_machine"}).Inc()
+
+	if len(*vm.StorageProfile.DataDisks) > 0 {
+		// There are disks attached hence need to detach them
+		vm.StorageProfile.DataDisks = &[]compute.DataDisk{}
+
+		_, errChan := vmClient.CreateOrUpdate(d.AzureMachineClass.Spec.ResourceGroup, machineID, vm, cancel)
```

Comment:
Can you please explain how does it work? 
Where are we making explicit detach calls to VM? , or does the empty-disk array instructs API to delete all disk?

Reply:
Yes hardik. VM with an update call with empty data disks is how detachment of data disks is done in Azure. They lack a proper SDK call for it. Refer here - https://github.com/Azure/azure-sdk-for-go/issues/1638#issuecomment-384733952